src/io/stun.rs

```diff
@@ -360,177 +364,137 @@ impl Method {
     }
 }
 
-#[derive(Clone, PartialEq, Eq)]
-pub enum Attribute<'a> {
-    MappedAddress,                // TODO
-    Username(&'a str),            // < 128 utf8 chars
-    MessageIntegrity(&'a [u8]),   // 20 bytes sha-1
-    MessageIntegrityMark,         // 20 bytes sha-1
-    ErrorCode(u16, &'a str),      // 300-699 and reason phrase < 128 utf8 chars
-    UnknownAttributes,            // TODO
-    Realm(&'a str),               // < 128 utf8 chars
-    Nonce(&'a str),               // < 128 utf8 chars
-    XorMappedAddress(SocketAddr), // 0x0020
-    Software(&'a str),
-    AlternateServer,  // TODO
-    Fingerprint(u32), // crc32
-    FingerprintMark,  // crc32
+#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
+pub struct Attributes<'a> {
+    username: Option<&'a str>,              // < 128 utf8 chars
+    message_integrity: Option<&'a [u8]>,    // 20 bytes sha-1
+    error_code: Option<(u16, &'a str)>,     // 300-699 and reason phrase < 128 utf8 chars
+    realm: Option<&'a str>,                 // < 128 utf8 chars
+    nonce: Option<&'a str>,                 // < 128 utf8 chars
+    xor_mapped_address: Option<SocketAddr>, // 0x0020
+    software: Option<&'a str>,
```

Comment:
If we add a doc comment to `software`, I think cargo fmt will align all these comments by itself.

Reply:
I'll see what I can do!